Avoid extracting the top errors in a ConcurrentMap instead use a List

reda-idtaleb opened the pull request.

reda-idtaleb committed:
Change the method ErrorsMap#collectTopXErrors() to return a sorted list of ErrorTypeInfo instances + change the affected files by this modif

reda-idtaleb committed:
move the method toOpenMetric() to the ErrorTypeInfo class

reda-idtaleb committed:
fix a bug in the front: the top errors metrics are remove from the datasets each log period

We connot register all the top errors in the datsets, because these metrics change every log period. Otherwise, this can lead to `undefined`` exception raised by the top-errors componenent. It can access the metrics of an old top error that is not available in the current log.

reda-idtaleb committed:
Add a first test for topErrors feature.

reda-idtaleb committed:
add some docs in ErrorsMap

reda-idtaleb committed:
add equals() and hash() method in ErrorTypeInfo class

reda-idtaleb committed:
add tests for top errors in MicrometerRegistryTest:

reda-idtaleb committed:
add a test to verify that the top errors list is sorted following the
occurrences of the errors

reda-idtaleb committed:
remove unused imports

pmouawad commented on ulp-observability-front/src/main/front/src/app/components/ulp-observability-dashboard/ulp-observability-dashboard.component.ts:
```diff
@@ -154,7 +154,12 @@ export class UlpObservabilityDashboardComponent implements OnInit{
+      // Checks if the key doesn't start with ``errorEveryPeriods_``. The metrics of the top errors 
+      // always starts with "errorEveryPeriods_", they should be removed before the next pushing 
+      // of data. This is necessary because the top errors recieved change with each logging period.
+      if (!key.startsWith("errorEveryPeriods_")) { 
```
Hello @reda-idtaleb , I am not sure to understand between code and comment, I see that if key.startsWith("errorEveryPeriods_" , this won't be called:
> nextDataset[key] = {};

So why do you write "they should be removed before the next pushing"

reda-idtaleb replied:
The errors shouldn't be copied in the new datasets. Before this modification, all the keys (from the current datsets) are copied into the new datatsets, but it shouldn't be the case for the errors. Then, we set the current datasets to the new datasets.

pmouawad commented on back/src/test/java/com/ubikloadpack/jmeter/ulp/observability/server/ULPObservabilityMetricServletTest.java:
```diff
@@ -115,6 +120,27 @@ public void whenTotalLabelCouldContainsSpacesExpectMetricsGeneratedAndLogged() t
 	
+	@Test
+	@DisplayName("When a sample fails due to an error, expect the occurred error is reported as a top error")
+	public void whenErrorOccurredOnSampleExpectItOnTopErrors() throws Exception {
```
Can you add a test, so that we have at least 5 different type of error with different occurences and check if result is fine ?

reda-idtaleb committed:
add a new test for top erros when there are at least 5 different errors

reda-idtaleb committed:
add an integration test to test the case when a sample fails with a
success code then the associated error type will be an `Assertion
failed`

reda-idtaleb committed:
tests refactoring

reda-idtaleb committed:
Add an other test for top errors feature

- The test verifies the computation of the error rate among errors and
the error rate among requests.
- code Refactoring

reda-idtaleb committed:
add a test to test the openmetric format when there are topRrrors

reda-idtaleb commented:
Hello, 
This PR can be merged. I added all the necessary tests for top error feature.
